File: backend/app/schemas/backtest.py
from __future__ import annotations

from datetime import date
from typing import List, Optional

from pydantic import BaseModel, Field, HttpUrl


class BacktestItemSchema(BaseModel):
    code: str
    name: str
    buy_date: date
    buy_price: float
    sell_date: date
    sell_price: float
    ret: float
    excess: float
    ann: float
    sharpe: Optional[float] = None
    mdd: Optional[float] = None
    calmar: Optional[float] = None
    score: Optional[float] = None
    grade: Optional[str] = None
    flags: List[str] = Field(default_factory=list)


class EquityPoint(BaseModel):
    date: date
    portfolio_nv: float
    bench_nv: float


class BacktestSummary(BaseModel):
    win_rate: float
    ret: float
    ann: float
    bench_ret: float
    bench_ann: float
    excess: float
    sharpe: Optional[float] = None
    mdd: Optional[float] = None
    calmar: Optional[float] = None


class BacktestWindow(BaseModel):
    start: date
    end: date
    trading_days: int


class BacktestResponse(BaseModel):
    bt_id: str
    window: BacktestWindow
    benchmark: str
    summary: BacktestSummary
    equity: List[EquityPoint]
    items: List[BacktestItemSchema]


class BacktestRequest(BaseModel):
    stocks: List[str]
    recommend_date: date
    end_date: Optional[date] = None
    benchmark: str = "HS300"
    price_adjust: str = "post"


class BacktestListResponse(BaseModel):
    items: List[BacktestResponse]

File: frontend/src/types/backtest.ts
export interface BacktestWindow {
  start: string | null;
  end: string | null;
  trading_days: number;
}

export interface BacktestSummary {
  win_rate?: number;
  ret?: number;
  ann?: number;
  bench_ret?: number;
  bench_ann?: number;
  excess?: number;
  sharpe?: number;
  mdd?: number;
  calmar?: number;
  grade?: string;
}

export interface BacktestItem {
  code: string;
  name: string;
  buy_date: string;
  buy_price?: number;
  sell_date?: string;
  sell_price?: number;
  ret?: number;
  excess?: number;
  ann?: number;
  sharpe?: number;
  mdd?: number;
  calmar?: number;
  score?: number;
  grade?: string;
  flags?: string[];
}

export interface ItemEquityPoint {
  date: string;
  ret: number;
}

export interface ItemEquitySeries {
  code: string;
  name: string;
  points: ItemEquityPoint[];
}

export interface BacktestResponse {
  bt_id: string;
  window: BacktestWindow;
  benchmark: string;
  summary: BacktestSummary;
  equity: Array<{ date: string; portfolio_nv: number; bench_nv: number }>;
  item_equities: ItemEquitySeries[];
  items: BacktestItem[];
}
